File: java-extensions/hadoop-ext/src/main/java/com/starrocks/connector/hadoop/KerberosLogin.java
// Copyright 2021-present StarRocks, Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.starrocks.connector.hadoop;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.security.SecurityUtil;
import org.apache.hadoop.security.UserGroupInformation;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.net.InetAddress;

/**
 * Logs the JVM embedded in BE/CN in to Kerberos from a keytab and keeps the ticket fresh.
 * libhdfs and the JNI connectors share this JVM, so they all inherit the login user.
 * Called from C++ at startup, only when both the principal and the keytab are configured.
 */
public class KerberosLogin {
    private static final Logger LOGGER = LoggerFactory.getLogger(KerberosLogin.class);

    private static final String AUTHENTICATION_KEY = "hadoop.security.authentication";
    private static final String KERBEROS_AUTHENTICATION = "kerberos";
    private static final String RELOGIN_THREAD_NAME = "kerberos-relogin";

    private static boolean loggedIn = false;

    private KerberosLogin() {
    }

    /**
     * @throws IOException if the login fails; the caller aborts startup rather than run with
     *                     a half-configured security context
     */
    public static synchronized void login(String principal, String keytab, int reloginIntervalSecond)
            throws IOException {
        if (loggedIn) {
            return;
        }
        if (principal == null || principal.isEmpty() || keytab == null || keytab.isEmpty()) {
            throw new IOException("kerberos_principal and kerberos_keytab must both be set");
        }
        File keytabFile = new File(keytab);
        if (!keytabFile.isFile() || !keytabFile.canRead()) {
            throw new IOException("kerberos_keytab is not a readable file: " + keytab);
        }

        String resolvedPrincipal =
                SecurityUtil.getServerPrincipal(principal, InetAddress.getLocalHost().getCanonicalHostName());

        // Keeps whatever core-site.xml on the classpath provides (auth_to_local rules and such)
        // and only forces the authentication method.
        Configuration conf = new Configuration();
        conf.set(AUTHENTICATION_KEY, KERBEROS_AUTHENTICATION);
        UserGroupInformation.setConfiguration(conf);
        UserGroupInformation.loginUserFromKeytab(resolvedPrincipal, keytab);
        loggedIn = true;

        LOGGER.info("{} Kerberos login succeeded, principal: {}, keytab: {}",
                HadoopExt.LOGGER_MESSAGE_PREFIX, resolvedPrincipal, keytab);

        startReloginThread(reloginIntervalSecond);
    }

    private static void startReloginThread(int reloginIntervalSecond) {
        long intervalMs = reloginIntervalSecond * 1000L;
        Thread thread = new Thread(() -> {
            while (true) {
                try {
                    Thread.sleep(intervalMs);
                    UserGroupInformation.getLoginUser().checkTGTAndReloginFromKeytab();
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    return;
                } catch (Throwable e) {
                    LOGGER.warn("{} failed to renew the Kerberos ticket, keeping the current one",
                            HadoopExt.LOGGER_MESSAGE_PREFIX, e);
                }
            }
        }, RELOGIN_THREAD_NAME);
        thread.setDaemon(true);
        thread.start();
    }
}

File: be/src/util/kerberos_login.cpp
// Copyright 2021-present StarRocks, Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "util/kerberos_login.h"

#include "common/config.h"
#include "jni.h"
#include "udf/java/java_udf.h"
#include "udf/java/utils.h"
#include "util/defer_op.h"

#define CHECK_JNI_EXCEPTION(env, message)                                                          \
    if (jthrowable thr = env->ExceptionOccurred(); thr) {                                          \
        std::string jni_error_message = JVMFunctionHelper::getInstance().dumpExceptionString(thr); \
        env->ExceptionClear();                                                                     \
        env->DeleteLocalRef(thr);                                                                  \
        return Status::InternalError(fmt::format("{}, error: {}", message, jni_error_message));    \
    }

namespace starrocks {

static const char* kKerberosLoginClassName = "com/starrocks/connector/hadoop/KerberosLogin";
static const char* kLoginMethodSignature = "(Ljava/lang/String;Ljava/lang/String;I)V";

static Status do_kerberos_login() {
    RETURN_IF_ERROR(detect_java_runtime());

    // check JNIEnv before calling JVMFunctionHelper::getInstance() to avoid crash
    if (getJNIEnv() == nullptr) {
        return Status::InternalError("get JNIEnv failed");
    }
    JNIEnv* env = JVMFunctionHelper::getInstance().getEnv();

    jclass login_cls = env->FindClass(kKerberosLoginClassName);
    CHECK_JNI_EXCEPTION(env, "find class KerberosLogin failed")
    LOCAL_REF_GUARD_ENV(env, login_cls);

    jmethodID login_method = env->GetStaticMethodID(login_cls, "login", kLoginMethodSignature);
    CHECK_JNI_EXCEPTION(env, "get method login failed")

    jstring principal = env->NewStringUTF(config::kerberos_principal.c_str());
    LOCAL_REF_GUARD_ENV(env, principal);
    jstring keytab = env->NewStringUTF(config::kerberos_keytab.c_str());
    LOCAL_REF_GUARD_ENV(env, keytab);

    env->CallStaticVoidMethod(login_cls, login_method, principal, keytab,
                              static_cast<jint>(config::kerberos_relogin_check_interval_second));
    CHECK_JNI_EXCEPTION(env, "kerberos login failed")

    return Status::OK();
}

bool is_kerberos_login_configured() {
    return !config::kerberos_principal.empty() && !config::kerberos_keytab.empty();
}

Status login_kerberos_if_configured() {
    if (config::kerberos_principal.empty() && config::kerberos_keytab.empty()) {
        return Status::OK();
    }
    if (config::kerberos_principal.empty() || config::kerberos_keytab.empty()) {
        return Status::InvalidArgument("kerberos_principal and kerberos_keytab must both be set");
    }
    return call_hdfs_scan_function_in_pthread(do_kerberos_login)->get_future().get();
}

} // namespace starrocks
